ui/src/dashboards/components/InfluxQEditor.tsx

```diff
@@ -0,0 +1,554 @@
+// Libraries
+import React, {Component} from 'react'
+import _ from 'lodash'
+import classnames from 'classnames'
+import {
+  Controlled as ReactCodeMirror,
+  IInstance as CMEditor,
+} from 'react-codemirror2'
+
+// Components
+import TemplateDrawer from 'src/shared/components/TemplateDrawer'
+import QueryStatus from 'src/shared/components/QueryStatus'
+import {ErrorHandling} from 'src/shared/decorators/errors'
+import OnClickOutside from 'src/shared/components/OnClickOutside'
+
+// Utilis
+import {getDeep} from 'src/utils/wrappers'
+
+// APIs
+import {replace as replaceQueryTemplates} from 'src/shared/apis/query'
+
+// Constants
+import {
+  MATCH_INCOMPLETE_TEMPLATES,
+  applyMasks,
+  insertTempVar,
+  unMask,
+} from 'src/tempVars/constants'
+
+// Types
+import {Template, QueryConfig} from 'src/types'
+import {EditorChange, Position} from 'codemirror'
+
+interface Selection {
+  start: Position
+  end: Position
+}
+
+interface TempVar {
+  tempVar: string
+}
+
+interface State {
+  focused: boolean
+  editedQueryText: string
+  templatingQueryText: string
+  isTemplating: boolean
+  selectedTemplate: TempVar
+  isShowingTemplateValues: boolean
+  filteredTemplates: Template[]
+  isSubmitted: boolean
+}
+
+interface Props {
+  query: string
+  onUpdate: (text: string) => void
+  config: QueryConfig
+  isInDataExplorer?: boolean
+  templates: Template[]
+  onClickOutside: () => void
+}
+
+const CODE_MIRROR_OPTIONS = {
+  tabIndex: 1,
+  mode: 'influxQL',
+  readonly: false,
+  lineNumbers: false,
+  autoRefresh: true,
+  theme: 'influxql',
+  completeSingle: false,
+  lineWrapping: true,
+}
+
+const TEMPLATE_START = ':'
+
+const NOOP = () => {}
+const FIRST_TEMP_VAR = '0.tempVar'
```

Comment:
this is nice, though i think it obscures why `getDeep` was necessary above. what do you think about calling it something like `FIRST_TEMP_VAR_NESTED` to preserve that clarity in its consumption?